browser/modules/SearchProvider.jsm

```diff
@@ -0,0 +1,265 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+/*globals Services, XPCOMUtils, Task, SearchSuggestionController, PrivateBrowsingUtils, FormHistory*/
+
+"use strict";
+
+this.EXPORTED_SYMBOLS = [
+  "SearchProvider",
+];
+
+const {
+  classes: Cc,
+  interfaces: Ci,
+  utils: Cu
+} = Components;
+
+Cu.import("resource://gre/modules/Services.jsm");
+Cu.import("resource://gre/modules/Promise.jsm");
+Cu.import("resource://gre/modules/Task.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+Cu.importGlobalProperties(['URL']);
+
+XPCOMUtils.defineLazyModuleGetter(this, "FormHistory",
+  "resource://gre/modules/FormHistory.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "PrivateBrowsingUtils",
+  "resource://gre/modules/PrivateBrowsingUtils.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "SearchSuggestionController",
+  "resource://gre/modules/SearchSuggestionController.jsm");
+
+const stringBundle = Services.strings.createBundle("chrome://global/locale/autocomplete.properties");
+const MAX_LOCAL_SUGGESTIONS = 3;
+const MAX_SUGGESTIONS = 6;
+
+this.SearchProvider = {
+  // This is used to handle search suggestions.  It maps xul:browsers to objects
+  // { controller, previousFormHistoryResult }.  See getSuggestions().
+  _suggestionMap: new WeakMap(),
+
+  get searchSuggestionUIStrings() {
+    if (this._searchSuggestionUIStrings) {
+      return this._searchSuggestionUIStrings;
+    }
+    this._searchSuggestionUIStrings = {};
+    let searchBundle = Services.strings.createBundle("chrome://browser/locale/search.properties");
+    let stringNames = ["searchHeader", "searchPlaceholder", "searchForKeywordsWith",
+      "searchWithHeader", "searchSettings"
+    ];
+    for (let name of stringNames) {
+      this._searchSuggestionUIStrings[name] = searchBundle.GetStringFromName(name);
+    }
+    return this._searchSuggestionUIStrings;
+  },
+
+  get state() {
+    var t = Task.async(function* () {
+      let state = {
+        engines: [],
+        currentEngine: yield this.getCurrentEngine(),
+      };
+      let pref = Services.prefs.getCharPref("browser.search.hiddenOneOffs");
+      let hiddenList = pref ? pref.split(",") : [];
+      for (let engine of Services.search.getVisibleEngines()) {
+        if (hiddenList.indexOf(engine.name) !== -1) {
+          continue;
+        }
+        let uri = engine.getIconURLBySize(16, 16);
+        state.engines.push({
+          name: engine.name,
+          iconBuffer: yield this._arrayBufferFromDataURI(uri),
+        });
+      }
+      return state;
+    }.bind(this))();
+
+    return t;
+  },
+
+  performSearch: function(browser, data) {
+    this._ensureDataHasProperties(data, [
```

Comment:
this should be designed like this (if we event want this... this is alien to JS):


```JS
function assureProps(requiredProperties){
    return function(data){
       return requiredProperties.every(prop => prop in data);
   }
}

const assureSearch = assureProps("engineName",
      "searchString",
     "healthReportKey",
     "searchPurpose", 
]);
// then later... 
assureSearch(data);
```